File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int l = 0, r = 0, ans = 0;

        Map<Character, Integer> mp = new HashMap<>();
        
        while (r < n) {
            char currentChar = s.charAt(r);

            if (!mp.containsKey(currentChar) || mp.get(currentChar) < l) {
                ans = Math.max(ans, r - l + 1);
                mp.put(currentChar, r);
            } else {
                l = mp.get(currentChar) + 1;
                mp.put(currentChar, r);
                ans = Math.max(ans, r - l + 1);
            }
            r++;
        }
        
        return ans;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int i = 0, j = 0;
        int result = 0;
        unordered_map<char, int> mp;

        while(j < n){
            mp[s[j]]++;

            while(mp[s[j]] > 1){
                mp[s[i]]--;
                i++;
            }

            result = max(result, j - i + 1);
            
            j++;
        }

        return result;
    }
};